File: src/main/resources/static/insumo/InsumoService.js
function insumoService($http, APP_CONFIG) {
	
	var rest_url = APP_CONFIG.REST_BASE_URL;
	
	this.findAll = function() {
		return $http.get(rest_url + '/insumos');
	};

	this.findAllPage = function(page, size) {
		var pageParams = "?page="+page+"&size="+size;
		return $http.get(rest_url + '/insumos/page' + pageParams);
	};
	
	this.findAllPageFilterBy = function(filtro, page, size) {
		var pageParams = "?page="+page+"&size="+size;
		return $http.post(rest_url + '/insumos/page/filterBy' + pageParams, filtro);
	};
	
	this.save = function(insumo) {
		return $http.post(rest_url + '/insumos', insumo);
	};
	
	this.del = function(idInsumo) {
		return $http.delete(rest_url + '/insumos/' + idInsumo);
	};
	
};

var depends = [
	'$http',
	'APP_CONFIG',
	insumoService  
];

app.service('insumoService', depends);

File: src/main/java/com/foundation/service/InsumoService.java
package com.foundation.service;

import java.util.List;

import org.apache.commons.collections4.CollectionUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.context.annotation.RequestScope;

import com.foundation.dao.InsumoDAO;
import com.foundation.filtroConsulta.FiltroConsultaInsumo;
import com.foundation.model.Composicao;
import com.foundation.model.Insumo;
import com.foundation.utils.CustomCollectionUtils;
import com.foundation.validador.ValidadorInsumoBuilder;

@Service
@RequestScope
public class InsumoService extends AbstractService {

	@Autowired
	private InsumoDAO insumoDAO;
	
	@Autowired
	private ComposicaoService composicaoService;
	
	@Transactional(propagation = Propagation.NOT_SUPPORTED)
	public Page<Insumo> findAll(Pageable page) {
		return insumoDAO.findAll(page);
	}
	
	@Transactional(propagation = Propagation.NOT_SUPPORTED)
	public Page<Insumo> findByFilter(FiltroConsultaInsumo filtro, Pageable page) {
		return insumoDAO.findByFilter(filtro, page);
	}
	
	public Insumo save(Insumo insumo) {
		ValidadorInsumoBuilder.newInstance()
			.validarCampoObrigatorio("nome", insumo.getNome())
			.validarCampoObrigatorio("quantidade", insumo.getQuantidade())
			.validarCampoObrigatorio("medida", insumo.getMedida())
			.assertValid();
		return insumoDAO.save(insumo);
	}

	@Transactional(propagation = Propagation.NOT_SUPPORTED)
	public List<Insumo> findAll() {
		return CustomCollectionUtils.toList(insumoDAO.findAll());
	}

	public void delete(Long id) {
		ValidadorInsumoBuilder.newInstance()
			.validarExisteProdutoCompostoPorInsumo(id, this)
			.assertValid();
		insumoDAO.delete(id);
	}

	public Insumo findOne(Long id) {
		return insumoDAO.findOne(id);
	}
	
	public boolean existeProdutoCompostoPeloInsumo(Long idInsumo) {
		List<Composicao> composicoesAssociadas = composicaoService.findAllByInsumo(idInsumo);
		return CollectionUtils.isNotEmpty(composicoesAssociadas);
	}
	
}